docs/models.md

```diff
@@ -1,198 +1,645 @@
+- Augments the prompt with additional guidance about the desired output format.
+  This also has the side effect of specifying to the model what content exactly
+  you want to generate (for example, not only suggest a menu item but also
+  generate a description, a list of allergens, and so on).
+- Parses the model output into a JavaScript object.
+- Verifies that the output conforms with the schema.
 
-## How to generate content
+To get structured output from a successful generate call, use the response
+object's `output()` method:
 
-`generate` is a helper function for working with models.
+```ts
+type MenuItem = z.infer<typeof MenuItemSchema>;
 
-To just call the model:
+const output: MenuItem | null = llmResponse.output();
+```
 
-```javascript
-import { generate } from '@genkit-ai/ai';
-import { gemini15Flash } from '@genkit-ai/vertexai';
+#### Handling errors
 
-(async () => {
+Note in the prior example that the output method can return `null`. This can
+happen when the model fails to generate output that conforms to the schema. You
+can also detect this condition by catching the `NoValidCandidatesError`
+exception thrown by generate:
+
+```ts
+import { NoValidCandidatesError } from "@genkit-ai/ai";
+```
+
+```ts
+try {
+  llmResponse = await generate(/* ... */);
+} catch (e) {
+  if (e instanceof NoValidCandidatesError) {
+    // Output doesn't conform to schema.
+  }
+}
+```
+
+The best strategy for dealing with such errors will depend on your exact use
+case, but here are some general hints:
+
+- **Try a different model**. For structured output to succeed, the model must be
+  capable of generating output in JSON. The most powerful LLMs, like Gemini and
+  Claude, are versatile enough to do this; however, smaller models, such as some
+  of the local models you would use with Ollama, might not be able to generate
+  structured output reliably unless they have been specifically trained to do
+  so.
+
+- **Make use of Zod's coercion abilities**: You can specify in your schemas that
+  Zod should try to coerce non-conforming types into the type specified by the
+  schema. If your schema includes primitive types other than strings, using Zod
+  coercion can reduce the number of `generate()` failures you experience. The
+  following version of MenuItemSchema uses type conversion to automatically
+  correct situations where the model generates calorie information as a string
+  instead of a number:
+
+  ```ts
+  const MenuItemSchema = z.object({
+    name: z.string(),
+    description: z.string(),
+    calories: z.coerce.number(),
+    allergens: z.array(z.string()),
+  });
+  ```
+
+- **Retry the generate() call**. If the model you've chosen only rarely fails to
+  generate conformant output, you can treat the error as you would treat a
+  network error, and simply retry the request using some kind of incremental
+  back-off strategy.
+
+- **Generate more candidates**. Another strategy for dealing with intermittent
```

Comment:
We're going to get rid of `candidates` in the 0.6 release in favor of calling `generate()` multiple times in parallel as the end result is the same and not all models support `candidates`.